core/src/pollers/poll_buffer.rs

```diff
@@ -16,102 +17,110 @@ use temporal_sdk_core_protos::temporal::api::workflowservice::v1::{
     /// Called every time the number of pollers is changed
     num_pollers_changed: Option<Box<dyn Fn(usize) + Send + Sync>>,
     active_pollers: Arc<AtomicUsize>,
 }
 
-struct ActiveCounter<'a>(&'a AtomicUsize);
-impl<'a> ActiveCounter<'a> {
-    fn new(a: &'a AtomicUsize) -> Self {
-        a.fetch_add(1, Ordering::Relaxed);
-        Self(a)
+struct ActiveCounter<'a, F: Fn(usize)>(&'a AtomicUsize, Option<F>);
+impl<'a, F> ActiveCounter<'a, F>
+where
+    F: Fn(usize),
+{
+    fn new(a: &'a AtomicUsize, change_fn: Option<F>) -> Self {
+        let v = a.fetch_add(1, Ordering::Relaxed) + 1;
+        if let Some(cfn) = change_fn.as_ref() {
+            cfn(v);
+        }
+        Self(a, change_fn)
     }
 }
-impl Drop for ActiveCounter<'_> {
+impl<F> Drop for ActiveCounter<'_, F>
+where
+    F: Fn(usize),
+{
     fn drop(&mut self) {
-        self.0.fetch_sub(1, Ordering::Relaxed);
+        let v = self.0.fetch_sub(1, Ordering::Relaxed) - 1;
+        if let Some(cfn) = self.1.as_ref() {
+            cfn(v)
+        }
     }
 }
 
 impl<T> LongPollBuffer<T>
 where
     T: Send + Debug + 'static,
 {
-    pub fn new<FT>(
+    pub(crate) fn new<FT>(
         poll_fn: impl Fn() -> FT + Send + Sync + 'static,
+        poll_semaphore: Arc<MeteredSemaphore>,
         max_pollers: usize,
-        buffer_size: usize,
         shutdown: CancellationToken,
+        num_pollers_handler: Option<impl Fn(usize) + Send + Sync + 'static>,
     ) -> Self
     where
         FT: Future<Output = pollers::Result<T>> + Send,
     {
-        let (tx, rx) = channel(buffer_size);
-        let polls_requested = Arc::new(Semaphore::new(0));
+        let (tx, rx) = unbounded_channel();
         let active_pollers = Arc::new(AtomicUsize::new(0));
         let join_handles = FuturesUnordered::new();
         let pf = Arc::new(poll_fn);
+        let nph = num_pollers_handler.map(Arc::new);
         for _ in 0..max_pollers {
             let tx = tx.clone();
             let pf = pf.clone();
             let shutdown = shutdown.clone();
-            let polls_requested = polls_requested.clone();
             let ap = active_pollers.clone();
+            let poll_semaphore = poll_semaphore.clone();
+            let nph = nph.clone();
             let jh = tokio::spawn(async move {
+                let nph = nph.as_ref().map(|a| a.as_ref());
                 loop {
                     if shutdown.is_cancelled() {
                         break;
                     }
-                    let sp = tokio::select! {
-                        sp = polls_requested.acquire() => sp.expect("Polls semaphore not dropped"),
+                    let permit = tokio::select! {
+                        p = poll_semaphore.acquire_owned() => p,
                         _ = shutdown.cancelled() => continue,
                     };
-                    let _active_guard = ActiveCounter::new(ap.as_ref());
+                    let permit = if let Ok(p) = permit {
```

Comment:
When is permit not `Ok`? Is it only on shutdown?
If that's the case you might be able to remove the check for `shutdown.is_cancelled()` at the top of this loop.

Reply:
It'd be exceedingly hard to reason about since that will only happen when the semaphore is dropped. Given it's refcounted all over the place I don't want to have to think about that.